Add packed virtqueue support

prakash-shekhar opened the pull request:
### Summary of the PR

Adds support for packed virtqueues, implementing the VirtIO packed virtqueue format alongside the existing split virtqueue support. Packed virtqueues offer improved performance through better cache locality and reduced memory overhead compared to traditional split virtqueues.

 Changes:
  1. VringConfigData structure (src/backend.rs):
    - Updated documentation to clearly distinguish between split and packed virtqueue address semantics
    - For packed virtqueues: desc_table_addr becomes the packed descriptor ring, used_ring_addr becomes driver event suppression, avail_ring_addr becomes device event suppression
  2. vhost-user protocol support (src/vhost_user/message.rs):
    - Added RING_PACKED feature flag (bit 34, value 0x4_0000_0000)
    - Added VHOST_VRING_F_PACKED vring flag (bit 1, value 0x2)
    - Implemented packed virtqueue inflight I/O structures (DescStatePacked, QueueRegionPacked)
  3. vhost-kern support (src/vhost_kern/mod.rs):
    - Extended kernel backend to support packed virtqueue address configuration
    - Added handling of packed layout in vring setup
  4. Test coverage:
    - Added test_packed_virtqueue_features() to validate feature flag definitions
    - Added test_packed_vring_addr_validation() to test packed virtqueue address validation/alignment

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR have Signed-Off-By trailers (with
  `git commit -s`), and the commit message has max 60 characters for the
  summary and max 75 characters for each description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

prakash-shekhar commented on vhost/src/vhost_kern/mod.rs:
```diff
@@ -73,26 +76,63 @@ pub trait VhostKernBackend: AsRawFd {
+        #[cfg(feature = "vhost-user")]
+        let is_packed = config_data.flags & VhostUserVringAddrFlags::VHOST_VRING_F_PACKED.bits() != 0;
+        #[cfg(not(feature = "vhost-user"))]
+        let is_packed = false;
```
Yep you're right vhost-kern should work independently, I'll fix that.

prakash-shekhar commented on vhost/src/vhost_kern/mod.rs:
```diff
@@ -79,7 +79,8 @@ pub trait VhostKernBackend: AsRawFd {
         // Check if packed ring format is being used
         #[cfg(feature = "vhost-user")]
-        let is_packed = config_data.flags & VhostUserVringAddrFlags::VHOST_VRING_F_PACKED.bits() != 0;
+        let is_packed =
```
Can you clarify what squash means, like squashing commits?

prakash-shekhar commented on vhost/CHANGELOG.md:
```diff
@@ -3,6 +3,7 @@
 ## [Unreleased]
 
 ### Added
+- Add support for packed virtqueues in vhost-user and vhost-kern backends
```
Understood, will fix.

germag commented on vhost-user-backend/Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 name = "vhost-user-backend"
-version = "0.18.0"
+version = "0.19.0"
 authors = ["The Cloud Hypervisor Authors"]
```
@stefano-garzarella is it ok to add a feature and create a release i the same PR?

stefano-garzarella replied:
nope, but I think this come up from merging `main` in this branch, indeed I asked to rebase it on main instead of merging, otherwise review it's a mess

germag commented on vhost/src/vhost_user/message.rs:
```diff
@@ -777,12 +782,28 @@ impl VhostUserMsgValidator for VhostUserVringAddr {
+        if self.flags & VhostUserVringAddrFlags::VHOST_VRING_F_PACKED.bits() != 0 {
+            // Packed ring validation:
+            // - descriptor ring must be 16-byte aligned
+            if self.descriptor & 0xf != 0 {
+                return false;
+            }
+            // - driver/device event suppression areas must be 4-byte aligned
+            if self.available & 0x3 != 0 || self.used & 0x3 != 0 {
+                return false;
+            }
+        } else {
+            // Split ring validation (current logic)
+            if self.descriptor & 0xf != 0 {
+                return false;
+            } else if self.available & 0x1 != 0 {
+                return false;
+            } else if self.used & 0x3 != 0 {
+                return false;
+            }
```
this is almost duplicated code, the only difference is the available's alignment, could this be simplified?

prakash-shekhar commented:
Hey thank you all for the feedback. I appreciate all the comments, I'm still trying to get better at this. I won't make these mistakes again. I consolidated everything into one commit, rebased, and fixed the issues you mentioned. I'd appreciate any further feedback as well.
